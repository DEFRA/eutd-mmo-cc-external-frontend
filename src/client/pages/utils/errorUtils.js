import i18n from '../../../i18n';
import enTranslations from '../../../locales/uk/en.json';

export function toGovukErrors(obj) {
  return Object.keys( obj).map( o => {
    return {targetName: o, text: obj[o], key: o, message: (obj[o] in enTranslations) ? i18n.t(obj[o]) : obj[o]};
  });
}

export function onHandleErrorClick(targetName, speciesRow, vesselIndex) {
  let elem;

  if (speciesRow !== undefined && vesselIndex !== undefined) {
    elem = document.getElementById(targetName + '-' + speciesRow + '-' + vesselIndex);
  } else {
    elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  }

  if (elem) {
    elem.scrollIntoView({ behavior: 'smooth' });
    elem.focus({ preventScroll: true});
  }
}

export function onHandleErrorClickWithoutFocus(targetName) {
  const elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  elem.scrollIntoView({ behavior: 'smooth' });
}

export function scrollToErrorIsland() {
  setTimeout( () => {
    document.getElementById('errorIsland').scrollIntoView({behavior: 'smooth'});
    document.getElementById('errorIsland').focus({ preventScroll: true});
  },100);
}

export function scrollToField(id) {
  setTimeout( () => {
    document.getElementById(id).scrollIntoView({behavior: 'smooth'});
    document.getElementById(id).focus({ preventScroll: true});
  },100);
}

export function scrollToFieldName(name) {
  setTimeout( () => {
    document.getElementsByName(name)[0].scrollIntoView({behavior: 'smooth'});
    document.getElementsByName(name)[0].focus({ preventScroll: true});
  },100);
}

