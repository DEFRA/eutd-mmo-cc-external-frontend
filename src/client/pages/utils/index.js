import React, {Fragment} from 'react';

export * from './errorUtils';


export function formatAddress( ...adresssLines) {
  return adresssLines.filter( a => a ).map( (a, index) => <Fragment key={index}>{a}<br/></Fragment> );
}

export function constructAddress(addressArray) {
  let address = '';
  if (addressArray.length === 4) {
    if (addressArray[0]) {
      address += addressArray[0];
    }
    if (addressArray[0] && addressArray[1]) {
      address += ' ';
    }
    if (addressArray[1]) {
      address += addressArray[1];
    }
    if (addressArray[2]) {
      address += ', ' + addressArray[2];
    }
    if (addressArray[3]) {
      address += '. ' + addressArray[3];
    }
  }
  return address;
}

export const hasAdminOverride = (exportPayload) => {
  const isLandingOverridden = (landing) =>
    landing &&
    landing.model &&
    landing.model.vessel &&
    landing.model.vessel.vesselOverriddenByAdmin;

  return exportPayload &&
    exportPayload.items &&
    exportPayload.items.some(
      (item) =>
        item.landings &&
        item.landings.some((landing) => isLandingOverridden(landing))
    )
    ? true
    : false;
};