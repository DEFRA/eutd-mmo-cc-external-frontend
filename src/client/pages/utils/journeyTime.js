export function startJourney() {
  window.startJourney = new Date(Date.now());
}

export function journeyTime() {
  if( !window.startJourney) return 0;
  const timeInSeconds = (new Date(Date.now()) - window.startJourney)/1000;
  return timeInSeconds.toFixed(0);
}
