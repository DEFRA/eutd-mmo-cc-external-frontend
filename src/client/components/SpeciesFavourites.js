import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'lodash';
import { GridCol } from 'govuk-react';
import FavouritesAutocomplete from './FavouritesAutocomplete';
import { scrollToErrorIsland } from '../pages/utils';
import { withTranslation } from 'react-i18next';
import { t } from 'i18next';
import ManageYourProductFavouritesLink from './ManageYourProductFavouritesLink';
export class SpeciesFavourites extends Component {
  constructor(props) {
    super(props);
    this.state = this.init(props.selectedSpeciesFavourite);
  }

  init(initialState) {
    if (!isEmpty(initialState)) {
      return initialState;
    }

    return ({
      commodity_code: '',
      commodity_code_description: '',
      id: '',
      selectedState: '',
      selectedStateLabel: '',
      selectedCommodityCode: '',
      selectedCommodityCodeLabel: '',
      selectedPresentation: '',
      selectedPresentationLabel: '',
      species: '',
      speciesCode: '',
      scientificName: '',
      selectedFavourite: '',
    });
  }

  setSpecies =  (favouriteObj) => {
    if (favouriteObj) {
      this.setState({
        selectedFavourite: `${favouriteObj.species} ${favouriteObj.stateLabel}, ${favouriteObj.presentationLabel}, ${favouriteObj.commodity_code}`,
        ...favouriteObj,
      }, () => {
        this.props.setSelectedSpeciesFavourite(this.state);
      });
    } else {
      this.setState({
        species: '',
        selectedFavourite: '',
      });
    }
  };

  addButtonHandler = async () => {
    try {
      const selectedData = {
        btn_submit: '',
        isFavourite: true,
        redirect:
          '/create-catch-certificate/:documentNumber/what-are-you-exporting',
        ...this.state,
      };

      const response = await this.props.addSpeciesPerUser(
        selectedData,
        this.props.documentNumber
      );

      if (response.payload) {
        this.setState({selectedFavourite: ''});
      }
    } catch (e) {
      scrollToErrorIsland();
    }
  };


  render() {
    const { products, errors } = this.props;
    return (
      <>
        <div id="0" className="adding-species">
          <div className="properties-group input-species-block">
            <FavouritesAutocomplete
              label={t('ccFavouritesPageProductTableHeaderTwo')}
              defaultValue={this.state.selectedFavourite}
              id="product"
              name="product"
              error={errors}
              onChange={this.setSpecies}
              htmlFor={'product'}
              allFavourites={products}
            />
          </div>
          <ManageYourProductFavouritesLink />
          <GridCol>
            <button
              id="submit"
              name="submit"
              className="button"
              style={{ marginLeft: '-0.8em ', marginTop: '1.5em' }}
              onClick={this.addButtonHandler}
            >
              {t('ccSpeciesFavouriteButtonText')}
            </button>
          </GridCol>
        </div>
      </>
    );
  }
}


SpeciesFavourites.propTypes = {
  documentNumber: PropTypes.string,
  addSpeciesPerUser: PropTypes.func,
  errors: PropTypes.object,
  products: PropTypes.array,
  selectedSpeciesFavourite: PropTypes.object,
  setSelectedSpeciesFavourite: PropTypes.func
};

export default withTranslation() (SpeciesFavourites);