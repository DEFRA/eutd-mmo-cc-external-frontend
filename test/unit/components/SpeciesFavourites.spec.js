import React from 'react';
import { shallow, mount } from 'enzyme';
import SpeciesFavourites from '../../../src/client/components/SpeciesFavourites';
import { act } from 'react-dom/test-utils';
import * as ErrorUtils from '../../../src/client/pages/utils/errorUtils';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

const mockAddSpeciesPerUser = jest.fn().mockResolvedValue({
  payload: {},
});
const mockShowSelectedFavourite = jest.fn();

const history = createMemoryHistory();

const getWrapper = (fullDepth = false, products = [], errors = {},) => {
  return fullDepth
    ? mount(
        <Router history={history}>
          <SpeciesFavourites
            addSpeciesPerUser={mockAddSpeciesPerUser}
            errors={errors}
            products={products}
            setSelectedSpeciesFavourite={mockShowSelectedFavourite}
            selectedSpeciesFavourite={undefined}
          />
        </Router>
      )
    : shallow(
        <Router history={history}>
          <SpeciesFavourites
            addSpeciesPerUser={mockAddSpeciesPerUser}
            errors={errors}
            products={products}
            setSelectedSpeciesFavourite={mockShowSelectedFavourite}
            selectedSpeciesFavourite={undefined}
          />
        </Router>
      );
};

describe('SpeciesFavourites', () => {
  it('should mount successfully', () => {
    const wrapper = getWrapper();
    expect(wrapper).toBeDefined();
  });

  it('should generate snapshot for the Plane Details page', () => {
    const wrapper = getWrapper(true);
    const { container } = render(wrapper)
    
    expect(container).toMatchSnapshot();
  })

  it('should load with an initial state', () => {
    const wrapper = getWrapper(true);
    expect(wrapper.find('SpeciesFavourites').state()).toEqual({
      commodity_code: '',
      commodity_code_description: '',
      id: '',
      selectedState: '',
      selectedStateLabel: '',
      selectedCommodityCode: '',
      selectedCommodityCodeLabel: '',
      selectedFavourite: '',
      selectedPresentation: '',
      selectedPresentationLabel: '',
      species: '',
      speciesCode: '',
      scientificName: '',
    });
  });

  it('should have an add product button', () => {
    const wrapper = getWrapper(true);
    const addBtn = wrapper.find('button#submit');
    expect(addBtn.exists()).toBeTruthy();
  });

  it('should add selected favourite to the products', () => {
    const wrapper = getWrapper(true);
    const addBtn = wrapper.find('button#submit');
    addBtn.simulate('click', {});

    expect(mockAddSpeciesPerUser).toHaveBeenCalled();
  });

  it('should set the correct state', () => {
    const initialState = {
      commodity_code: '',
      commodity_code_description: '',
      id: '',
      selectedState: '',
      selectedStateLabel: '',
      selectedCommodityCode: '',
      selectedCommodityCodeLabel: '',
      selectedFavourite:'',
      selectedPresentation: '',
      selectedPresentationLabel: '',
      species: '',
      speciesCode: '',
      scientificName: '',
    };

    const wrapper = mount(
      <Router history={history}>
        <SpeciesFavourites
          addSpeciesPerUser={mockAddSpeciesPerUser}
          errors={{}}
          products={[]}
          setSelectedSpeciesFavourite={mockShowSelectedFavourite}
          selectedSpeciesFavourite={undefined}
        />
      </Router>
    );

    const component = wrapper.find('SpeciesFavourites');
    component.setState(initialState);

    const favouritesInput = wrapper.find('FavouritesAutocomplete');

    const expectedState = {
      commodity_code: 'some-code',
      commodity_code_description: '',
      id: '',
      selectedState: '',
      selectedStateLabel: 'state',
      selectedCommodityCode: '',
      selectedCommodityCodeLabel: '',
      selectedPresentation: '',
      selectedFavourite: 'some-specie state presentation some-code',
      selectedPresentationLabel: 'presentation',
      species: 'some-specie',
      speciesCode: '',
      scientificName: '',
    };

    act(() => {
      favouritesInput.prop('onChange')(expectedState);
    });
    expect(mockShowSelectedFavourite).toHaveBeenCalledWith(expectedState);
    expect(component.state()).toEqual(expectedState);
  });

  it('should assigned the default value for favouritesInput if selectedFavourite is empty' ,() => {

    const wrapper = mount(
      <Router history={history}>
        <SpeciesFavourites
          addSpeciesPerUser={mockAddSpeciesPerUser}
          errors={{}}
          products={[]}
          showSelectedFavourite= {mockShowSelectedFavourite}
          selectedSpeciesFavourite={{ selectedFavourite: 'some-specie state presentation some-cod'}}
        />
      </Router>
    );

    expect(wrapper.find('FavouritesAutocomplete').props().defaultValue).toEqual('some-specie state presentation some-cod');
  });

  it('should call ScrollToErrorIsland if an error is thrown', () => {
    const wrapper = mount(
      <Router history={history}>
        <SpeciesFavourites
          addSpeciesPerUser={mockAddSpeciesPerUser}
          errors={[{ product: 'Select a product favourite from the list' }]}
          products={[]}
          showSelectedFavourite ={mockShowSelectedFavourite}
          selectedSpeciesFavourite=''
        />
      </Router>
    );

    const mockScrollToErrorIsland = jest.spyOn(
      ErrorUtils,
      'scrollToErrorIsland'
    );

    mockAddSpeciesPerUser.mockImplementation(() => {
      throw 'error';
    });
    const addBtn = wrapper.find('button#submit');
    addBtn.simulate('click', {});

    expect(mockScrollToErrorIsland).toHaveBeenCalled();
  });

  it('should render product errors if any', () => {
    const errors = {
      errors: [
        {
          targetName: 'product',
          text: 'Select a product favourite from the list',
        },
      ],
      productError: 'Select a product favourite from the list',
    };

    const wrapper = getWrapper(true, [], errors);

    expect(wrapper.find('FavouritesAutocomplete').props().error).toEqual(
      errors
    );
  });
});

describe('setSpecies', () => {
  it('Should set the selection to none when user clears the input', () => {
    const wrapper = getWrapper(true);
    const component = wrapper.find('SpeciesFavourites');
    const favouritesInput = wrapper.find('FavouritesAutocomplete');

    act(() => {
      favouritesInput.prop('onChange')();
    });
    expect(component.state().species).toEqual('');
  });
});
