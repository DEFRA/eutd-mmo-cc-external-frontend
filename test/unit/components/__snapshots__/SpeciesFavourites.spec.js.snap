// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeciesFavourites should generate snapshot for the Plane Details page 1`] = `
<div>
  <div
    class="adding-species"
    id="0"
  >
    <div
      class="properties-group input-species-block"
    >
      <label
        class="e1vcssft0 css-137pbp-StyledLabel-StyledHoc e4u1u2g0"
        for="product"
        id="label-product"
      >
        <div
          id="favouritesLabel"
        >
          Product
        </div>
        <div
          class="autocomplete__wrapper"
        >
          <div
            style="border: 0px; height: 1px; margin-bottom: -1px; margin-right: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
          >
            <div
              aria-atomic="true"
              aria-live="polite"
              id="product__status--A"
              role="status"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              id="product__status--B"
              role="status"
            />
          </div>
          
          <input
            aria-activedescendant="false"
            aria-autocomplete="list"
            aria-describedby="product__assistiveHint"
            aria-expanded="false"
            aria-owns="product__listbox"
            autocomplete="off"
            class="autocomplete__input autocomplete__input--default"
            id="product"
            name="product"
            placeholder=""
            role="combobox"
            type="text"
            value=""
          />
          <ul
            class="autocomplete__menu autocomplete__menu--overlay autocomplete__menu--hidden"
            id="product__listbox"
            role="listbox"
          />
          <span
            id="product__assistiveHint"
            style="display: none;"
          >
            When autocomplete results are available use up and down arrows to review and enter to select.  Touch device users, explore by touch or with swipe gestures.
          </span>
        </div>
      </label>
    </div>
    <a
      href="/manage-favourites"
      id="manageYourProductFavourites"
    >
      Manage your product favourites
    </a>
    <div
      class="css-1w80l5h-StyledColumn e4u1u2g0"
    >
      <button
        class="button"
        id="submit"
        name="submit"
        style="margin-left: -0.8em; margin-top: 1.5em;"
      >
        Add Product
      </button>
    </div>
  </div>
</div>
`;
